README.md
# Rollup multiple .scss, .sass and .css imports

<a href="LICENSE">
  <img src="https://img.shields.io/badge/license-MIT-brightgreen.svg" alt="Software License" />
</a>
<a href="https://github.com/thgh/rollup-plugin-scss/issues">
  <img src="https://img.shields.io/github/issues/thgh/rollup-plugin-scss.svg" alt="Issues" />
</a>
<a href="http://standardjs.com/">
  <img src="https://img.shields.io/badge/code%20style-standard-brightgreen.svg" alt="JavaScript Style Guide" />
</a>
<a href="https://npmjs.org/package/rollup-plugin-scss">
  <img src="https://img.shields.io/npm/v/rollup-plugin-scss.svg?style=flat-squar" alt="NPM" />
</a>
<a href="https://github.com/thgh/rollup-plugin-scss/releases">
  <img src="https://img.shields.io/github/release/thgh/rollup-plugin-scss.svg" alt="Latest Version" />
</a>

## Installation
```
# Rollup v0.60+ and v1+
npm install --save-dev rollup-plugin-scss

# Rollup v0.59 and below
npm install --save-dev rollup-plugin-scss@0
```

## Usage
```js
// rollup.config.js
import scss from 'rollup-plugin-scss'

export default {
  input: 'input.js',
  output: {
    file: 'output.js',
    format: 'esm'
  },
  plugins: [
    scss() // will output compiled styles to output.css
  ]
}
```

```js
// entry.js
import './reset.scss'
```

### Options

Options are passed to the sass compiler ([node-sass] by default). Refer to [ the Sass docs](https://sass-lang.com/documentation/js-api#options) for more details on these options. <br/>
One notable option is `indentedSyntax` which you'll need if you're parsing Sass syntax instead of Scss syntax. (e.g. when extracting a Vue `<style lang="sass">` tag) <br/>
By default the plugin will base the filename for the css on the bundle destination.

```js
scss({
  // Choose *one* of these possible "output:..." options
  // Default behaviour is to write all styles to the bundle destination where .js is replaced by .css
  output: true,

  // Filename to write all styles to
  output: 'bundle.css',

  // Callback that will be called ongenerate with two arguments:
  // - styles: the contents of all style tags combined: 'body { color: green }'
  // - styleNodes: an array of style objects: { filename: 'body { ... }' }
  output: function (styles, styleNodes) {
    writeFileSync('bundle.css', styles)
  },

  // Disable any style output or callbacks, import as string
  output: false,
  
  // An array of paths to resolve @import from (default: ["node_modules"]);
  // in the context of a monorepo you might want to include more paths.
  includePaths: ["node_modules", "../../node_modules"],
  
  // Choose files to include in processing (default: ['/**/*.css', '/**/*.scss', '/**/*.sass'])
  include: [],
  
  // Choose files to exclude from processing, (default: undefined) 
  exclude: [],

  // Determine if node process should be terminated on error (default: false)
  failOnError: true,

  // Prefix global scss. Useful for variables and mixins.
  prefix: `@import "./fonts.scss";`,

  // Use a node-sass compatible compiler (default: node-sass)
  sass: require('sass'),

  // Run postcss processor before output
  processor: css => postcss([autoprefixer({ overrideBrowserslist: "Edge 18" })]),

  // Process resulting CSS
  processor: css => css.replace('/*date*/', '/* ' + new Date().toJSON() + ' */'),

  // Add file/folder to be monitored in watch mode so that changes to these files will trigger rebuilds.
  // Do not choose a directory where rollup output or dest is pointed to as this will cause an infinite loop
  watch: 'src/styles/components',
  watch: ['src/styles/components', 'src/multiple/folders'],
})
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Contributions and feedback are very welcome. New features should include a test.

To get it running:
  1. Clone the project.
  2. `npm install`

## Credits

- [Thomas Ghysels](https://github.com/thgh)
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.

[link-author]: https://github.com/thgh
[link-contributors]: ../../contributors
[rollup-plugin-vue]: https://www.npmjs.com/package/rollup-plugin-vue
[rollup-plugin-buble]: https://www.npmjs.com/package/rollup-plugin-buble
[rollup-plugin-babel]: https://www.npmjs.com/package/rollup-plugin-babel
[node-sass]: https://www.npmjs.com/package/node-sass
[sass]: https://www.npmjs.com/package/sass


CHANGELOG.md
# Changelog

All notable changes to `rollup-plugin-scss` will be documented in this file.

## [Unreleased]

## [3.0.0] - 2021-06-29

### Added

- Add insert option @syJSdev
- Add `sourceMap` option to enable generation of source map @astappiev
- Add Yarn PnP support` @eagerestwolf
- Refactor to Typescript
- Automated testing using Github Actions

### Updated

- Prefer sass over node-sass
- A `processor` can receive map as second parameter and return `{ css: string, map?: string }`
- Remove `node-sass` from optionalDependencies @astappiev <br/>
  **You have to specify `node-sass` or `sass` in your project dependencies alongside `rollup-plugin-scss`**

## [2.6.1] - 2020-10-01

### Updated

- Move node-sass to optionalDependencies @weizhenye

## [2.6.0] - 2020-08-14

### Fixed

- Resolve processor as a promise during transform step @geotrev

### Added

- Add support for postcss processor

## [2.5.0] - 2020-05-07

### Updated

- Fix includePaths before processing @mach25

## [2.4.0] - 2020-04-13

### Added

- Add `sass` option @riri

## [2.2.0] - 2020-04-11

### Added

- Add `watch` option @JimSchofield

## [2.1.0] - 2019-12-22

### Added

- Add `prefix` option @jackprosser

## [2.0.0] - 2019-12-22

### Changed

- Add `node_modules/` in includePaths by default
- Fix cases where output `css` is null or undefined
- Update dependencies

## [1.0.0] - 2019-02-04

### Update

- Update `ongenerate` to `generateBundle`

[unreleased]: https://github.com/thgh/rollup-plugin-scss/compare/v2.2.0...HEAD
[2.2.0]: https://github.com/thgh/rollup-plugin-scss/compare/v2.1.0...v2.2.0
[2.1.0]: https://github.com/thgh/rollup-plugin-scss/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/thgh/rollup-plugin-scss/compare/v1.0.0...v2.0.0
[1.0.0]: https://github.com/thgh/rollup-plugin-scss/compare/v0.0.1...v1.0.0
[0.0.1]: https://github.com/thgh/rollup-plugin-scss/releases


test/entry.js
import './lib.scss'


test/input.js
import './input.scss'

console.log('scss imported')


test/output.js
console.log('scss imported');


index.d.ts
import { Processor } from 'postcss'
import { Plugin } from 'rollup'

/**
 * Callback that will be called ongenerate with two arguments.
 *
 * ---
 * **styles**
 *
 * The contents of all style tags combined.
 *
 * > _eg:_ `body { color: green }`
 *
 * ---
 *
 * **styleNodes**
 *
 * An array of style objects.
 *
 * >  _eg:_ `{ filename: 'body { ... }' }`
 */
declare type IOutput = (
  styles?: string,
  styleNodes?: {
    [filename: string]: string
  }
) => void

export interface SCSSPluginOptions {
  /**
   * Accepts either a boolean, string or function.
   * Choose *one* of these possible `output:...` options.
   * Default behaviour is to write all styles to the bundle destination
   * where `.js` is replaced by `.css.`
   *
   * ---
   *
   * Passing in boolean (default: true):
   *
   * ```js
   *
   * // Default behaviour will write to bundle destination.
   * { output: true }
   *
   * // Disable any style output or callbacks, import as string.
   * { output: false }
   * ```
   *
   * ---
   *
   * Passing in string (file path)
   *
   * ```js
   * {
   *   // Filename to write all styles to
   *   output: 'bundle.css'
   * }
   * ```
   *
   * ---
   *
   * Passing a function callback
   *
   * ```js
   * {
   *    // Callback that will be called ongenerate with two arguments
   *    output: function (styles, styleNodes) {
   *      writeFileSync('bundle.css', styles)
   *    }
   * }
   * ```
   * ---
   *
   * @default true
   */
  output?: boolean | string | IOutput,
  /**
   * Choose files to include in processing. Defaults to to all
   * .scss, .sass and .css files.
   */
  include?: string[],
  /**
   * Choose files to exclude from processing,
   *
   * @default undefined
   */
  exclude?: string[],
  /**
   * Determine if node process should be terminated on error.
   *
   * @default  false
   */
  failOnError?: boolean;
  /**
   * Prefix global scss. Useful for variables and mixins.
   *
   * @example `@import "./fonts.scss";`
   * @default ''
   */
  prefix?: string;
  /**
   * Use a `node-sass` compatible compiler.
   *
   * - [node-sass](https://www.npmjs.com/package/node-sass)
   * - [sass](https://www.npmjs.com/package/sass)
   *
   * ---
   *
   * @default require('node-sass')
   */
  sass?: any;

  /**
   * Process resulting CSS or run a `postcss` processor before output.
   *
   * @example
   *
   * // Running PostCSS Processor
   * {  processor: css => postcss([autoprefixer()]) }
   *
   * // Processing some resulting CSS
   * { processor: css => '.some-class { color: red;}' + css }
   */
  processor: (css?: string) => Processor

  /**
   * Add file/folder to be monitored in watch mode so
   * that changes to these files will trigger rebuilds.
   *
   * **IMPORTANT**
   * Do not choose a directory where rollup output or
   * dest is pointed to as this will cause an infinite loop.
   *
   * @example
   * { watch: 'src/styles/components' }
   *
   * // Array of folders
   * { watch: ['src/styles/components', 'src/multiple/folders'] }
   */
  watch?: string | string[],


  /**
   * Included paths, which are passed to node-sass.
   * In most cases, you can leave this option to execute its default,
   * which includes node_modules.
   *
   * @default  ['node_modules']
   */
  includePaths?: string[];

}

/**
 * Rollup multiple .scss, .sass and .css imports.
 */
export default function scss(options: SCSSPluginOptions): Plugin;


index.es.js
import { existsSync, mkdirSync } from 'fs'
import { dirname, basename } from 'path'
import { createFilter } from 'rollup-pluginutils'

/*
 * Create a style tag and append to head tag
 *
 * @param {String} css style
 * @return {String} css style
 */
function insertStyleFn(css) {
  if (!css) {
    return
  }
  if (typeof window === "undefined") {
    return
  }

  const style = document.createElement("style")

  style.setAttribute("type", "text/css")
  style.innerHTML = css
  document.head.appendChild(style)
  return css
}

export default function css (options = {}) {
  const filter = createFilter(options.include || ['/**/*.css', '/**/*.scss', '/**/*.sass'], options.exclude)
  let dest = options.output
  const insertStyleFnName = "___$insertStylesToHeader"

  const styles = {}
  const prefix = options.prefix ? options.prefix + '\n' : ''
  let includePaths = options.includePaths || ['node_modules/']
  includePaths.push(process.cwd())

  const compileToCSS = function (scss) {
    // Compile SASS to CSS
    if (scss.length) {
      includePaths = includePaths.filter((v, i, a) => a.indexOf(v) === i)
      try {
        const sass = options.sass || require('node-sass')
        const css = sass.renderSync(Object.assign({
          data: prefix + scss,
          includePaths
        }, options)).css.toString()
        // Possibly process CSS (e.g. by PostCSS)
        if (typeof options.processor === 'function') {
          const processor = options.processor(css, styles)

          // PostCSS support
          if (typeof processor.process === 'function') {
            return Promise.resolve(processor.process(css, { from: undefined }))
              .then(result => result.css)
          }

          return processor
        }
        return css
      } catch (e) {
        if (options.failOnError) {
          throw e
        }
        console.log()
        console.log(red('Error:\n\t' + e.message))
        if (e.message.includes('Invalid CSS')) {
          console.log(green('Solution:\n\t' + 'fix your Sass code'))
          console.log('Line:   ' + e.line)
          console.log('Column: ' + e.column)
        }
        if (e.message.includes('node-sass') && e.message.includes('find module')) {
          console.log(green('Solution:\n\t' + 'npm install --save node-sass'))
        }
        if (e.message.includes('node-sass') && e.message.includes('bindings')) {
          console.log(green('Solution:\n\t' + 'npm rebuild node-sass --force'))
        }
        console.log()
      }
    }
  }

  return {
    name: 'scss',
    intro() {
      if (options.insert === true) {
        return insertStyleFn.toString().replace(/insertStyleFn/, insertStyleFnName)
      }
    },
    transform (code, id) {
      if (!filter(id)) {
        return
      }

      // Add the include path before doing any processing
      includePaths.push(dirname(id))

      // Rebuild all scss files if anything happens to this folder
      // TODO: check if it's possible to get a list of all dependent scss files
      //       and only watch those
      if ('watch' in options) {
        const files = Array.isArray(options.watch) ? options.watch : [options.watch]
        files.forEach(file => this.addWatchFile(file))
      }

      if (options.insert === true) {
        // When the 'insert' is enabled, the stylesheet will be inserted into <head/> tag.
        return Promise.resolve(compileToCSS(code)).then((css) => ({
          code:
            "export default " + insertStyleFnName + "(" + JSON.stringify(css) + ")",
          map: { mappings: "" },
        }))
      } else if (options.output === false) {
        // When output is disabled, the stylesheet is exported as a string
        return Promise.resolve(compileToCSS(code)).then(css => ({
          code: 'export default ' + JSON.stringify(css),
          map: { mappings: '' }
        }))
      }

      // Map of every stylesheet
      styles[id] = code

      return ''
    },
    generateBundle (opts) {
      // No stylesheet needed
      if (options.output === false || options.insert === true) {
        return
      }

      // Combine all stylesheets
      let scss = ''
      for (const id in styles) {
        scss += styles[id] || ''
      }

      const css = compileToCSS(scss)

      // Resolve if processor returned a Promise
      Promise.resolve(css).then(css => {
        // Emit styles through callback
        if (typeof options.output === 'function') {
          options.output(css, styles)
          return
        }

        if (typeof css !== 'string') {
          return
        }

        if (typeof dest !== 'string') {
          // Don't create unwanted empty stylesheets
          if (!css.length) {
            return
          }

          // Guess destination filename
          dest = opts.dest || opts.file || 'bundle.js'
          if (dest.endsWith('.js')) {
            dest = dest.slice(0, -3)
          }
          dest = dest + '.css'
        }

        // Ensure that dest parent folders exist (create the missing ones)
        ensureParentDirsSync(dirname(dest))

        // Emit styles to file
        this.emitFile({
          type: 'asset',
          source: css,
          name: basename(dest),
          fileName: dest.replace(dirname(opts.dest || opts.file), '').replace(/^\\|\//, '')
        })
      })
    }
  }
}

function red (text) {
  return '\x1b[1m\x1b[31m' + text + '\x1b[0m'
}

function green (text) {
  return '\x1b[1m\x1b[32m' + text + '\x1b[0m'
}

function getSize (bytes) {
  return bytes < 10000
    ? bytes.toFixed(0) + ' B'
    : bytes < 1024000
      ? (bytes / 1024).toPrecision(3) + ' kB'
      : (bytes / 1024 / 1024).toPrecision(4) + ' MB'
}

function ensureParentDirsSync (dir) {
  if (existsSync(dir)) {
    return
  }

  try {
    mkdirSync(dir)
  } catch (err) {
    if (err.code === 'ENOENT') {
      ensureParentDirsSync(dirname(dir))
      ensureParentDirsSync(dir)
    }
  }
}


index.ts
import { existsSync, mkdirSync, writeFile } from 'fs'
import { basename, dirname, resolve } from 'path'
import { createFilter, CreateFilter } from 'rollup-pluginutils'

import type { Plugin } from 'rollup'

export interface CSSPluginOptions {
  exclude?: Parameters<CreateFilter>[1]
  failOnError?: boolean
  include?: Parameters<CreateFilter>[0]
  includePaths?: string[]
  insert?: boolean
  output?: string | false | ((css: string, styles: Styles) => void)
  prefix?: string
  processor?: (
    css: string,
    map: string,
    styles: Styles
  ) => CSS | Promise<CSS> | PostCSSProcessor
  sass?: SassRenderer
  sourceMap?: boolean
  verbose?: boolean
  watch?: string | string[]
}

type ImporterReturnType = { file: string } | { contents: string } | Error | null

type ImporterDoneCallback = (data: ImporterReturnType) => void

type CSS = string | { css: string; map: string }

interface MappedCSS {
  css: string
  map: string
}

interface Styles {
  [id: string]: string
}

interface PostCSSProcessor {
  process: (css: string, options?: any) => MappedCSS
}

interface SassRenderer {
  renderSync: (options: SassOptions) => SassResult
}

interface SassOptions {
  data: string
}

interface SassResult {
  css: Buffer
  map?: Buffer
}

export default function scss(options: CSSPluginOptions = {}): Plugin {
  const filter = createFilter(
    options.include || ['/**/*.css', '/**/*.scss', '/**/*.sass'],
    options.exclude
  )
  let dest = typeof options.output === 'string' ? options.output : null
  const insertStyleFnName = '___$insertStylesToHeader'

  const styles: Styles = {}
  const prefix = options.prefix ? options.prefix + '\n' : ''
  let includePaths = options.includePaths || ['node_modules/']
  includePaths.push(process.cwd())

  const compileToCSS = async function (scss: string) {
    // Compile SASS to CSS
    if (scss.length) {
      includePaths = includePaths.filter((v, i, a) => a.indexOf(v) === i)
      try {
        const sass = options.sass || loadSassLibrary()

        const render = sass.renderSync(
          Object.assign(
            {
              data: prefix + scss,
              outFile: dest,
              includePaths,
              importer: (
                url: string,
                prev: string,
                done: ImporterDoneCallback
              ): ImporterReturnType | void => {
                /* If a path begins with `.`, then it's a local import and this
                 * importer cannot handle it. This check covers both `.` and
                 * `..`.
                 *
                 * Additionally, if an import path begins with `url` or `http`,
                 * then it's a remote import, this importer also cannot handle
                 * that. */
                if (
                  url.startsWith('.') ||
                  url.startsWith('url') ||
                  url.startsWith('http')
                ) {
                  /* The importer returns `null` to defer processing the import
                   * back to the sass compiler. */
                  return null
                }

                /* If the requested path begins with a `~`, we remove it. This
                 * character is used by webpack-contrib's sass-loader to
                 * indicate the import is from the node_modules folder. Since
                 * this is so standard in the JS world, the importer supports
                 * it, by removing it and ignoring it. */
                const cleanUrl = url.startsWith('~')
                  ? url.replace('~', '')
                  : url

                /* Now, the importer uses `require.resolve()` to attempt
                 * to resolve the path to the requested file. In the case
                 * of a standard node_modules project, this will use Node's
                 * `require.resolve()`. In the case of a Plug 'n Play project,
                 * this will use the `require.resolve()` provided by the
                 * package manager.
                 * 
                 * This statement is surrounded by a try/catch block because
                 * if Node or the package manager cannot resolve the requested
                 * file, they will throw an error, so the importer needs to
                 * defer to sass, by returning `null`. 
                 * 
                 * The paths property tells `require.resolve()` where to begin
                 * resolution (i.e. who is requesting the file). */
                try {
                  const resolved = require.resolve(cleanUrl, {
                    paths: [prefix + scss]
                  })

                  /* Since `require.resolve()` will throw an error if a file
                   * doesn't exist. It's safe to assume the file exists and
                   * pass it off to the sass compiler. */
                  return { file: resolved }
                } catch (e: any) {
                  /* Just because `require.resolve()` couldn't find the file
                   * doesn't mean it doesn't exist. It may still be a local
                   * import that just doesn't list a relative path, so defer
                   * processing back to sass by returning `null` */
                  return null
                }
              }
            },
            options
          )
        )

        const css = render.css.toString()
        const map = render.map ? render.map.toString() : ''

        // Possibly process CSS (e.g. by PostCSS)
        if (typeof options.processor === 'function') {
          const result = await options.processor(css, map, styles)

          // TODO: figure out how to check for
          // @ts-ignore
          const postcss: PostCSSProcessor = result

          // PostCSS support
          if (typeof postcss.process === 'function') {
            return Promise.resolve(
              postcss.process(css, {
                from: undefined,
                to: dest,
                map: map ? { prev: map, inline: false } : null
              })
            )
          }

          // @ts-ignore
          const output: string | MappedCSS = result

          return stringToCSS(output)
        }
        return { css, map }
      } catch (e) {
        if (options.failOnError) {
          throw e
        }
        console.log()
        console.log(red('Error:\n\t' + e.message))
        if (e.message.includes('Invalid CSS')) {
          console.log(green('Solution:\n\t' + 'fix your Sass code'))
          console.log('Line:   ' + e.line)
          console.log('Column: ' + e.column)
        }
        if (e.message.includes('sass') && e.message.includes('find module')) {
          console.log(
            green(
              'Solution:\n\t' +
                'npm install --save-dev sass'
            )
          )
        }
        if (e.message.includes('node-sass') && e.message.includes('bindings')) {
          console.log(green('Solution:\n\t' + 'npm rebuild node-sass --force'))
        }
        console.log()
      }
    }
    return { css: '', map: '' }
  }

  return {
    name: 'scss',
    intro() {
      return options.insert === true
        ? insertStyleFn.replace(/insertStyleFn/, insertStyleFnName)
        : ''
    },
    async transform(code, id) {
      if (!filter(id)) {
        return
      }

      // Add the include path before doing any processing
      includePaths.push(dirname(id))

      // Rebuild all scss files if anything happens to this folder
      // TODO: check if it's possible to get a list of all dependent scss files
      //       and only watch those
      if (options.watch) {
        const files = Array.isArray(options.watch)
          ? options.watch
          : [options.watch]
        files.forEach(file => this.addWatchFile(file))
      }

      if (options.insert === true) {
        // When the 'insert' is enabled, the stylesheet will be inserted into <head/> tag.
        const { css, map } = await compileToCSS(code)
        return {
          code:
            'export default ' +
            insertStyleFnName +
            '(' +
            JSON.stringify(css) +
            ')',
          map: { mappings: '' }
        }
      } else if (options.output === false) {
        // When output is disabled, the stylesheet is exported as a string
        const { css, map } = await compileToCSS(code)
        return {
          code: 'export default ' + JSON.stringify(css),
          map: { mappings: '' }
        }
      }

      // Map of every stylesheet
      styles[id] = code

      return ''
    },
    async generateBundle(opts) {
      // No stylesheet needed
      if (options.output === false || options.insert === true) {
        return
      }

      // Combine all stylesheets
      let scss = ''
      for (const id in styles) {
        scss += styles[id] || ''
      }

      if (typeof dest !== 'string') {
        // Guess destination filename
        dest = opts.file || 'bundle.js'
        if (dest.endsWith('.js')) {
          dest = dest.slice(0, -3)
        }
        dest = dest + '.css'
      }

      const compiled = await compileToCSS(scss)

      if (typeof compiled !== 'object' || typeof compiled.css !== 'string') {
        return
      }

      // Emit styles through callback
      if (typeof options.output === 'function') {
        options.output(compiled.css, styles)
        return
      }

      // Don't create unwanted empty stylesheets
      if (!compiled.css.length) {
        return
      }

      // Ensure that dest parent folders exist (create the missing ones)
      ensureParentDirsSync(dirname(dest))

      // Emit styles to file
      writeFile(dest, compiled.css, err => {
        if (options.verbose !== false) {
          if (err) {
            console.error(red(err.toString()))
          } else if (compiled.css) {
            console.log(green(dest || '?'), getSize(compiled.css.length))
          }
        }
      })

      if (options.sourceMap && compiled.map) {
        let sourcemap = compiled.map
        if (typeof compiled.map.toString === 'function') {
          sourcemap = compiled.map.toString()
        }

        writeFile(dest + '.map', sourcemap, err => {
          if (options.verbose !== false && err) {
            console.error(red(err.toString()))
          }
        })
      }
    }
  }
}

/**
 * Create a style tag and append to head tag
 *
 * @param {String} css style
 * @return {String} css style
 */
const insertStyleFn = `function insertStyleFn(css) {
  if (!css) {
    return
  }
  if (typeof window === 'undefined') {
    return
  }

  const style = document.createElement('style');

  style.setAttribute('type', 'text/css');
  style.innerHTML = css;
  document.head.appendChild(style);
  return css
}`

function loadSassLibrary(): SassRenderer {
  try {
    return require('sass')
  } catch (e) {
    return require('node-sass')
  }
}

function stringToCSS(input: string | CSS): MappedCSS {
  if (typeof input === 'string') {
    return { css: input, map: '' }
  }
  return input
}

function red(text: string) {
  return '\x1b[1m\x1b[31m' + text + '\x1b[0m'
}

function green(text: string) {
  return '\x1b[1m\x1b[32m' + text + '\x1b[0m'
}

function getSize(bytes: number) {
  return bytes < 10000
    ? bytes.toFixed(0) + ' B'
    : bytes < 1024000
    ? (bytes / 1024).toPrecision(3) + ' kB'
    : (bytes / 1024 / 1024).toPrecision(4) + ' MB'
}

function ensureParentDirsSync(dir: string) {
  if (existsSync(dir)) {
    return
  }

  try {
    mkdirSync(dir)
  } catch (err) {
    if (err.code === 'ENOENT') {
      ensureParentDirsSync(dirname(dir))
      ensureParentDirsSync(dir)
    }
  }
}
